# Laboratorio 1 - Regresión Lineal

### Imports

In [125]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from ydata_profiling import ProfileReport

from importlib.metadata import version
print(f"Versión de Pandas: {version('pandas')}")
print(f"Versión de Matplotlib: {version('matplotlib')}")
print(f"Versión de Seaborn: {version('seaborn')}")

Versión de Pandas: 2.2.3
Versión de Matplotlib: 3.10.0
Versión de Seaborn: 0.13.2


### Carga de Archivos

In [126]:
datos_pacientes = pd.read_csv('./data/Datos Lab 1.csv', )
datos = datos_pacientes.copy()
datos.head()

,Patient ID,Date of Service,Sex,Age,Weight (kg),Height (m),BMI,Abdominal Circumference (cm),Blood Pressure (mmHg),Total Cholesterol (mg/dL),...,Physical Activity Level,Family History of CVD,Height (cm),Waist-to-Height Ratio,Systolic BP,Diastolic BP,Blood Pressure Category,Estimated LDL (mg/dL),CVD Risk Score,CVD Risk Level
0,isDx5313,"November 08, 2023",M,44.0,114.300,1.720,38.600,100.000,112/83,228.0,...,High,N,172.000,0.581,112.0,83.0,Hypertension Stage 1,121.0,19.880,HIGH
1,LHCK2961,20/03/2024,F,57.0,92.923,1.842,33.116,106.315,101/91,158.0,...,High,Y,184.172,0.577,101.0,91.0,Hypertension Stage 2,57.0,16.833,INTERMEDIARY
2,WjVn1699,2021-05-27,F,NaN,73.400,1.650,27.000,78.100,90/74,135.0,...,High,N,165.000,0.473,90.0,74.0,Normal,45.0,12.600,LOW
3,dCDO1109,"April 18, 2022",F,35.0,113.300,1.780,35.800,79.600,92/89,158.0,...,Moderate,Y,178.000,0.447,92.0,89.0,Hypertension Stage 1,94.0,14.920,HIGH
4,pnpE1080,01/11/2024,F,48.0,102.200,1.750,33.400,106.700,121/68,207.0,...,Low,Y,175.000,0.610,121.0,68.0,Elevated,128.0,18.870,HIGH


## 1. Exploración y Perfilamiento de los Datos

In [127]:
display(datos.sample(5))

,Patient ID,Date of Service,Sex,Age,Weight (kg),Height (m),BMI,Abdominal Circumference (cm),Blood Pressure (mmHg),Total Cholesterol (mg/dL),...,Physical Activity Level,Family History of CVD,Height (cm),Waist-to-Height Ratio,Systolic BP,Diastolic BP,Blood Pressure Category,Estimated LDL (mg/dL),CVD Risk Score,CVD Risk Level
1036,UQyn8657,30 Aug 24,F,37.0,15.036,1.870,25.900,87.60,113/75,252.0,...,High,Y,187.000,0.468,113.0,75.0,Normal,143.0,17.870,HIGH
1248,MeWu4854,04-13-2021,F,58.0,82.100,NaN,26.200,82.60,97/69,133.0,...,Low,N,177.000,0.467,97.0,69.0,Normal,60.0,12.750,HIGH
1390,ejaQ6145,"October 07, 2025",F,32.0,55.100,1.770,17.600,109.70,119/61,204.0,...,High,N,177.000,NaN,119.0,61.0,Normal,106.0,13.550,LOW
13,nepF9907,06/08/2020,M,74.0,51.858,1.608,33.294,71.58,99/78,124.0,...,Moderate,Y,160.768,0.445,99.0,NaN,Normal,34.0,16.089,LOW
946,ZIIY7203,27/09/2022,M,59.0,103.900,1.660,37.700,87.10,117/61,272.0,...,Low,N,166.000,0.525,117.0,61.0,Normal,211.0,18.830,HIGH


Hay varios formatos de los datos en la columna Date of Service.

In [128]:
datos.shape

(1639, 24)

In [129]:
display(datos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1639 entries, 0 to 1638
Data columns (total 24 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Patient ID                    1639 non-null   object 
 1   Date of Service               1639 non-null   object 
 2   Sex                           1639 non-null   object 
 3   Age                           1571 non-null   float64
 4   Weight (kg)                   1566 non-null   float64
 5   Height (m)                    1578 non-null   float64
 6   BMI                           1586 non-null   float64
 7   Abdominal Circumference (cm)  1578 non-null   float64
 8   Blood Pressure (mmHg)         1639 non-null   object 
 9   Total Cholesterol (mg/dL)     1571 non-null   float64
 10  HDL (mg/dL)                   1557 non-null   float64
 11  Fasting Blood Sugar (mg/dL)   1585 non-null   float64
 12  Smoking Status                1639 non-null   object 
 13  Dia

None

Hay 24 columnas con 1639 datos.

Las columnas con valores nulos son: Age,  Weight (kg), Height (m), BMI, Abdominal Circumference (cm), Total Cholesterol (mg/dL),  HDL (mg/dL), Fasting Blood Sugar (mg/dL),  Height (cm),  Waist-to-Height Ratio, Systolic BP, Diastolic BP, Estimated LDL (mg/dL) y CVD Risk Score

La columna con más valores nulos tiene el 94,8% de todos los datos

In [130]:
dict = pd.read_excel('./data/DiccPacientes.xlsx')
pd.set_option('display.max_colwidth', None)
dict

,Nombre Columna,Tipo de dato,Comentarios
0,Patient ID,String,Identificador del paciente
1,Date of Service,Date,Fecha de la atención
2,Sex,String,"Sexo (Femenino, Masculino)"
3,Age,Integer,Edad
4,Weight (kg),Float,Peso
5,Height (m),Float,Altura
6,BMI,Float,Índice de masa corporal
7,Abdominal Circumference (cm),Float,Circunferencia abdominal
8,Blood Pressure (mmHg),String,"Presión sanguínea, de la forma ""<Presión arterial sistólica>/<Presión arterial diastólica>"""
9,Total Cholesterol (mg/dL),Float,Colesterol total


Los tipos de datos del el diccionario y pandas están distintos, a excepeción de los floats

In [131]:
datos.describe()

,Age,Weight (kg),Height (m),BMI,Abdominal Circumference (cm),Total Cholesterol (mg/dL),HDL (mg/dL),Fasting Blood Sugar (mg/dL),Height (cm),Waist-to-Height Ratio,Systolic BP,Diastolic BP,Estimated LDL (mg/dL),CVD Risk Score
count,1571.000000,1566.000000,1578.000000,1586.000000,1578.000000,1571.000000,1557.000000,1585.000000,1571.000000,1563.000000,1578.000000,1554.000000,1582.000000,1610.000000
mean,46.803186,85.666006,1.757439,28.424744,91.538861,199.043673,56.183558,117.836860,175.770082,0.522440,125.632637,82.887536,113.235896,18.227281
std,13.039479,21.712504,0.118012,7.309275,13.427985,59.388670,16.721702,32.379634,11.695880,0.085692,22.577463,15.503625,61.435291,10.767666
min,6.134000,13.261000,1.371000,4.317000,49.542000,-1.256000,0.008000,15.306000,136.498000,0.250000,49.914000,31.720000,-92.055000,-20.057000
25%,37.000000,67.100000,1.666500,22.600000,79.700000,150.000000,42.000000,92.000000,167.000000,0.453000,108.000000,71.000000,62.000000,15.150000
50%,46.000000,86.314000,1.760000,28.000000,91.200000,199.000000,56.000000,115.000000,176.000000,0.519000,125.000000,82.000000,112.000000,16.967000
75%,55.000000,104.801500,1.850000,33.963000,102.267250,250.000000,70.000000,139.000000,185.000000,0.582000,141.000000,93.000000,159.000000,18.900000
max,89.420000,158.523000,2.146000,53.028000,136.336000,385.679000,110.315000,219.667000,214.394000,0.804000,202.711000,134.066000,317.314000,114.980000


Se identificaron múltiples problemas de calidad en los datos. En particular, se observaron valores imposibles físicamente como colesterol total y LDL negativos, así como puntajes de riesgo cardiovascular negativos, los cuales no tienen interpretación clínica directa. A su vez, es extraño tener un rango de edades tan grande.

También se evidencian outliers extremos en variables como presión arterial, altura y colesterol, que podrían influir de forma desproporcionada en modelos de regresión lineal.

In [132]:
((datos.isnull().sum()/datos.shape[0])).sort_values(ascending=False)

Diastolic BP                    0.051861
HDL (mg/dL)                     0.050031
Waist-to-Height Ratio           0.046370
Weight (kg)                     0.044539
Height (cm)                     0.041489
Age                             0.041489
Total Cholesterol (mg/dL)       0.041489
Height (m)                      0.037218
Abdominal Circumference (cm)    0.037218
Systolic BP                     0.037218
Estimated LDL (mg/dL)           0.034777
Fasting Blood Sugar (mg/dL)     0.032947
BMI                             0.032337
CVD Risk Score                  0.017694
Patient ID                      0.000000
Blood Pressure Category         0.000000
Smoking Status                  0.000000
Family History of CVD           0.000000
Physical Activity Level         0.000000
Diabetes Status                 0.000000
Date of Service                 0.000000
Blood Pressure (mmHg)           0.000000
Sex                             0.000000
CVD Risk Level                  0.000000
dtype: float64

In [133]:
nulos = datos[datos['Diastolic BP'].isna()].copy()

print(f"Filas con Diastolic BP nula: {nulos.shape[0]}")
indices = datos.index[datos['Diastolic BP'].isna()]
datos[datos['Diastolic BP'].isna()].copy().reset_index(drop=True)

Filas con Diastolic BP nula: 85


,Patient ID,Date of Service,Sex,Age,Weight (kg),Height (m),BMI,Abdominal Circumference (cm),Blood Pressure (mmHg),Total Cholesterol (mg/dL),...,Physical Activity Level,Family History of CVD,Height (cm),Waist-to-Height Ratio,Systolic BP,Diastolic BP,Blood Pressure Category,Estimated LDL (mg/dL),CVD Risk Score,CVD Risk Level
0,nepF9907,06/08/2020,M,74.0,51.858,1.608,33.294,71.58,99/78,124.0,...,Moderate,Y,160.768,0.445,99.0,NaN,Normal,34.0,16.089,LOW
1,AhYt1346,09-28-2020,M,41.0,71.300,1.730,23.800,107.90,139/61,253.0,...,Low,Y,173.000,NaN,139.0,NaN,Hypertension Stage 1,146.0,16.770,HIGH
2,fHqx3420,"July 09, 2024",M,55.0,70.200,1.720,23.700,70.00,121/68,300.0,...,High,Y,172.000,0.407,121.0,NaN,Elevated,193.0,17.064,HIGH
3,NGwd5164,2021-12-31,M,33.0,74.000,1.760,23.900,77.40,118/86,165.0,...,High,Y,176.000,0.440,118.0,NaN,Hypertension Stage 1,64.0,13.980,LOW
4,ICKt4668,2025-11-01,M,42.0,59.400,1.890,16.600,94.40,125/61,228.0,...,High,N,189.000,0.499,125.0,NaN,Elevated,127.0,32.120,INTERMEDIARY
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
80,cUUw7860,06-16-2023,F,37.0,75.400,1.800,23.300,86.50,149/78,296.0,...,High,N,NaN,0.481,149.0,NaN,Hypertension Stage 2,209.0,20.030,INTERMEDIARY
81,gynJ3543,25/01/2020,M,30.0,88.030,1.987,22.225,107.69,174/82,204.0,...,Moderate,Y,198.666,NaN,174.0,NaN,Hypertension Stage 1,124.0,19.225,HIGH
82,RsUs2064,2021-07-14,M,39.0,52.641,1.898,33.100,73.69,120/78,228.0,...,Moderate,Y,189.817,0.388,120.0,NaN,Elevated,160.0,19.180,INTERMEDIARY
83,ePpS4710,12/04/2022,M,48.0,50.100,1.770,16.000,104.10,146/95,210.0,...,Low,Y,177.000,0.588,146.0,NaN,Hypertension Stage 2,110.0,NaN,HIGH


In [134]:
nulos = datos[datos['Height (m)'].isna()].copy()

print(f"Filas con Height (m) nula: {nulos.shape[0]}")
indices = datos.index[datos['Height (m)'].isna()]
datos[datos['Height (m)'].isna()].copy().reset_index(drop=True)

Filas con Height (m) nula: 61


,Patient ID,Date of Service,Sex,Age,Weight (kg),Height (m),BMI,Abdominal Circumference (cm),Blood Pressure (mmHg),Total Cholesterol (mg/dL),...,Physical Activity Level,Family History of CVD,Height (cm),Waist-to-Height Ratio,Systolic BP,Diastolic BP,Blood Pressure Category,Estimated LDL (mg/dL),CVD Risk Score,CVD Risk Level
0,oJnb5472,05 Jun 25,M,34.0,102.600,NaN,NaN,96.900,137/81,240.0,...,High,N,139.265,0.587,137.0,81.0,Hypertension Stage 1,133.0,19.190,INTERMEDIARY
1,NBUl2902,01-21-2025,F,62.0,87.544,NaN,38.527,86.487,154/60,125.0,...,Moderate,Y,178.337,0.485,154.0,60.0,Hypertension Stage 2,47.0,19.905,HIGH
2,MmNI0458,"May 31, 2024",M,48.0,NaN,NaN,28.800,79.700,102/63,245.0,...,Low,Y,178.000,0.448,102.0,63.0,Normal,143.0,15.760,HIGH
3,XZCU4466,22/01/2024,F,52.0,97.800,NaN,31.600,94.800,124/92,216.0,...,High,N,176.000,0.539,124.0,92.0,Hypertension Stage 2,136.0,18.840,HIGH
4,SdtW2617,21/06/2021,M,NaN,115.400,NaN,36.800,91.900,142/64,NaN,...,High,N,177.000,0.519,142.0,64.0,Hypertension Stage 2,67.0,19.180,HIGH
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
56,inTX4760,12-10-2020,F,42.0,69.100,NaN,19.600,NaN,92/95,263.0,...,Low,N,188.000,0.545,92.0,95.0,Hypertension Stage 2,NaN,13.780,INTERMEDIARY
57,KeUy4784,01-23-2023,F,31.0,98.900,NaN,30.200,77.900,129/86,174.0,...,Low,Y,181.000,NaN,129.0,86.0,Hypertension Stage 1,107.0,15.970,HIGH
58,FMVb4279,06/03/2022,M,30.0,91.400,NaN,30.900,96.300,106/83,126.0,...,Moderate,N,172.000,0.560,106.0,83.0,Hypertension Stage 1,64.0,114.143,INTERMEDIARY
59,Qrlb4412,2025-05-02,M,73.0,118.888,NaN,35.129,97.390,101/95,147.0,...,Moderate,N,179.435,0.543,101.0,95.0,Hypertension Stage 2,74.0,17.016,LOW


Existen datos que tienen la altura en centimetros pero no en metros y viceversa

In [135]:
datos.duplicated(keep = False).sum()

np.int64(302)

In [136]:
# ids duplicados
dup_counts = (datos['Patient ID'].value_counts()
                        .loc[lambda s: s > 1]
                        .sort_values(ascending=False))
for id_, n in dup_counts.items():
    print(f"Id={id_} → {n} apariciones")

Id=THSJ8564 → 3 apariciones
Id=LdMm6225 → 3 apariciones
Id=lqUc7918 → 3 apariciones
Id=ChGR7779 → 3 apariciones
Id=szyX1885 → 3 apariciones
Id=PHii0023 → 3 apariciones
Id=svjU9851 → 3 apariciones
Id=sINw8855 → 3 apariciones
Id=EsZI4603 → 3 apariciones
Id=zZle5455 → 3 apariciones
Id=KQUs0708 → 3 apariciones
Id=HNpy8362 → 3 apariciones
Id=tlWh1188 → 3 apariciones
Id=VgtQ3260 → 3 apariciones
Id=HNwp6592 → 3 apariciones
Id=gdBF9655 → 3 apariciones
Id=ivTX4645 → 3 apariciones
Id=Xvbe7861 → 3 apariciones
Id=gXxm9399 → 3 apariciones
Id=TPHC2670 → 3 apariciones
Id=INdI1482 → 3 apariciones
Id=nmgP2712 → 3 apariciones
Id=FTEC4446 → 3 apariciones
Id=rfYk3516 → 3 apariciones
Id=RlsB8509 → 3 apariciones
Id=dJuC5084 → 3 apariciones
Id=STpP5810 → 3 apariciones
Id=GhAN9460 → 3 apariciones
Id=pEpZ9034 → 3 apariciones
Id=dSiv4949 → 3 apariciones
Id=cUUw7860 → 3 apariciones
Id=UIWC3599 → 3 apariciones
Id=ZDBx7052 → 3 apariciones
Id=fHqx3420 → 3 apariciones
Id=YLCe2926 → 3 apariciones
Id=kajW6905 → 3 apar

In [137]:
# Detalle de los registros
dup_table = (
    datos[datos['Patient ID'].duplicated(keep=False)]
    .copy()
    .assign(repeticiones=datos.groupby('Patient ID')['Patient ID'].transform('size'))
    .sort_values(['repeticiones', 'Patient ID'], ascending=[False, True])
)

# Mostrar tabla
dup_table

,Patient ID,Date of Service,Sex,Age,Weight (kg),Height (m),BMI,Abdominal Circumference (cm),Blood Pressure (mmHg),Total Cholesterol (mg/dL),...,Family History of CVD,Height (cm),Waist-to-Height Ratio,Systolic BP,Diastolic BP,Blood Pressure Category,Estimated LDL (mg/dL),CVD Risk Score,CVD Risk Level,repeticiones
17,AhYt1346,09-28-2020,M,41.0,71.300,1.730,23.800,107.9,139/61,253.0,...,Y,173.000,NaN,139.0,NaN,Hypertension Stage 1,146.0,16.770,HIGH,3
1227,AhYt1346,09-28-2020,M,41.0,71.300,1.730,23.800,107.9,139/61,253.0,...,Y,173.000,NaN,139.0,NaN,Hypertension Stage 1,146.0,-13.090,HIGH,3
1584,AhYt1346,09-28-2020,M,41.0,71.300,1.730,23.800,107.9,139/61,253.0,...,Y,173.000,NaN,139.0,NaN,Hypertension Stage 1,146.0,16.770,HIGH,3
130,BQvQ6431,09/11/2020,M,33.0,118.300,1.690,41.400,72.1,116/93,171.0,...,N,210.554,0.427,116.0,93.0,Hypertension Stage 2,97.0,17.500,LOW,3
1469,BQvQ6431,09/11/2020,M,33.0,118.300,1.690,41.400,72.1,116/93,171.0,...,N,210.554,0.427,116.0,93.0,Hypertension Stage 2,97.0,29.833,LOW,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
915,yvsn3005,28 Oct 20,F,60.0,54.300,1.810,16.600,99.1,133/65,187.0,...,N,181.000,0.548,133.0,65.0,Hypertension Stage 1,79.0,15.710,INTERMEDIARY,2
95,zcgB3048,12/06/2020,M,38.0,60.436,1.506,20.269,79.7,168/63,119.0,...,N,150.616,0.529,168.0,63.0,Hypertension Stage 2,43.0,14.834,HIGH,2
375,zcgB3048,12/06/2020,M,38.0,60.436,1.506,20.269,79.7,168/63,119.0,...,N,150.616,0.529,168.0,63.0,Hypertension Stage 2,43.0,14.834,HIGH,2
799,zxhX5525,"November 13, 2021",M,26.0,58.953,1.688,25.286,NaN,110/114,258.0,...,Y,168.763,0.455,110.0,114.0,Hypertension Stage 2,170.0,15.717,HIGH,2


In [138]:
datos['Date of Service'].value_counts()

Date of Service
09-20-2023           6
December 05, 2025    5
June 08, 2023        4
December 02, 2020    4
05-30-2025           4
                    ..
2024-09-24           1
11/09/2020           1
August 25, 2024      1
31/07/2022           1
2021-05-01           1
Name: count, Length: 1274, dtype: int64

### Hallazgos clave de la exploración
- Se identifican valores nulos principalmente en Distolic BP, HDL (mg/dL), Age, Weight, Height y BMI.
- Existen varios formatos para las fechas.
- Hay varios registros de pacientes repetidos.
- Los tipos de datos entre el Pandas y el diccionario no concuerdan
- Hay datos que si bien hacen sentido matemático, no se atienen a la lógica del negocio


## 2. Limpieza y Preparación

#### Corregir inconsistencia de alturas 

In [139]:
# Calcular Height (m) con BMI y Weight (kg) donde Height (m) es nulo
mask_height_nan = datos['Height (m)'].isna() & datos['Weight (kg)'].notna() & datos['BMI'].notna()
datos.loc[mask_height_nan, 'Height (m)'] = (datos.loc[mask_height_nan, 'Weight (kg)'] / datos.loc[mask_height_nan, 'BMI'])

# Caso 1: Height (m) vacío y Height (cm) disponible
mask_m_nan = datos['Height (m)'].isna() & datos['Height (cm)'].notna()
datos.loc[mask_m_nan, 'Height (m)'] = datos.loc[mask_m_nan, 'Height (cm)'] / 100

# Caso 2: Height (cm) vacío y Height (m) disponible
mask_cm_nan = datos['Height (cm)'].isna() & datos['Height (m)'].notna()
datos.loc[mask_cm_nan, 'Height (cm)'] = datos.loc[mask_cm_nan, 'Height (m)'] * 100

# Verificación rápida
datos[['Height (m)', 'Height (cm)']].isna().sum()


Height (m)     0
Height (cm)    0
dtype: int64

#### Calcular Weight (kg) con BMI y Height (kg)

In [140]:
mask_weight_nan = datos['Weight (kg)'].isna() & datos['BMI'].notna() & datos['Height (m)'].notna()
datos.loc[mask_weight_nan, 'Weight (kg)'] = datos.loc[mask_weight_nan, 'BMI'] * (datos.loc[mask_weight_nan, 'Height (m)'])

print(f"Valores nulos en Weight (kg) después de la imputación: {datos['Weight (kg)'].isna().sum()}")

Valores nulos en Weight (kg) después de la imputación: 1


In [141]:
datos['BMI'].isna().sum() & datos['Weight (kg)'].isna().sum()

np.int64(1)

In [142]:
datos_borrar = datos[datos['BMI'].isna() & datos['Weight (kg)'].isna()]
datos = datos.drop(datos_borrar.index)

datos['BMI'].isna().sum() & datos['Weight (kg)'].isna().sum()

np.int64(0)

#### Imputar CVD risk Score

In [143]:
median_cvd = datos['CVD Risk Score'].median()
datos['CVD Risk Score'].fillna(median_cvd, inplace=True)

datos['CVD Risk Score'].isna().sum()

C:\Users\scast\AppData\Local\Temp\ipykernel_23212\1584683367.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  datos['CVD Risk Score'].fillna(median_cvd, inplace=True)


np.int64(0)

#### Columnas con valores negativos

In [144]:
datos.loc[datos.loc[:, 'CVD Risk Score']<0]

,Patient ID,Date of Service,Sex,Age,Weight (kg),Height (m),BMI,Abdominal Circumference (cm),Blood Pressure (mmHg),Total Cholesterol (mg/dL),...,Physical Activity Level,Family History of CVD,Height (cm),Waist-to-Height Ratio,Systolic BP,Diastolic BP,Blood Pressure Category,Estimated LDL (mg/dL),CVD Risk Score,CVD Risk Level
154,LpyK7269,24/06/2024,F,26.0,57.250,1.859,29.971,81.215,111/76,147.0,...,High,N,185.864,0.437,111.0,76.0,Normal,60.0,-4.166,HIGH
232,nmgP2712,"October 15, 2022",M,44.0,97.800,1.890,27.400,91.900,129/61,166.0,...,Moderate,Y,189.000,0.486,129.0,61.0,Elevated,75.0,-2.319,HIGH
292,bfnb4412,"May 05, 2024",M,49.0,118.300,1.890,33.100,96.200,130/68,240.0,...,Moderate,N,189.000,0.509,130.0,68.0,Hypertension Stage 1,149.0,-18.977,INTERMEDIARY
513,NFgw6578,2022-04-06,F,37.0,75.200,1.800,23.200,NaN,108/67,187.0,...,Moderate,Y,180.000,0.598,108.0,67.0,Normal,108.0,-15.745,INTERMEDIARY
698,zpEJ0975,"January 01, 2023",F,40.0,115.200,1.660,41.800,75.200,119/72,300.0,...,Moderate,Y,166.000,0.453,119.0,72.0,Normal,230.0,-20.057,INTERMEDIARY
738,yaTS5266,10-19-2021,M,60.0,78.900,1.850,23.100,101.500,94/78,133.0,...,Low,Y,185.000,0.549,94.0,78.0,Normal,64.0,-18.859,HIGH
1053,qfwo1163,2024-10-04,F,57.0,89.651,1.603,27.783,70.403,129/108,NaN,...,Moderate,Y,160.326,0.439,129.0,108.0,Hypertension Stage 2,112.0,-15.968,INTERMEDIARY
1132,Keaq9838,04 Sep 22,M,44.0,108.300,1.800,33.400,96.600,140/83,134.0,...,High,Y,180.000,0.537,140.0,83.0,Hypertension Stage 1,58.0,-9.629,INTERMEDIARY
1227,AhYt1346,09-28-2020,M,41.0,71.300,1.730,23.800,107.900,139/61,253.0,...,Low,Y,173.000,NaN,139.0,NaN,Hypertension Stage 1,146.0,-13.090,HIGH


In [145]:
datos.loc[datos.loc[:, 'Estimated LDL (mg/dL)']<0]

,Patient ID,Date of Service,Sex,Age,Weight (kg),Height (m),BMI,Abdominal Circumference (cm),Blood Pressure (mmHg),Total Cholesterol (mg/dL),...,Physical Activity Level,Family History of CVD,Height (cm),Waist-to-Height Ratio,Systolic BP,Diastolic BP,Blood Pressure Category,Estimated LDL (mg/dL),CVD Risk Score,CVD Risk Level
69,nOqf9878,28 Jul 23,F,60.0,54.800,1.770000,17.500,81.200,108/81,108.0,...,Moderate,N,177.000,0.459,108.0,81.0,Hypertension Stage 1,-2.000,11.060,INTERMEDIARY
176,RLth4450,23/07/2022,M,48.0,81.400,1.630000,30.600,106.700,149/81,191.0,...,Moderate,N,163.000,0.655,149.0,NaN,Hypertension Stage 1,-86.972,19.390,HIGH
236,vzlz0512,2024-08-05,F,54.0,74.241,1.555000,24.554,116.941,104/107,105.0,...,Moderate,N,155.545,0.752,104.0,107.0,Hypertension Stage 2,-10.000,12.211,HIGH
301,DkCD0536,2025-03-26,M,49.0,83.065,1.971000,28.154,92.423,103/81,104.0,...,Low,Y,197.127,0.469,103.0,81.0,Hypertension Stage 1,-15.000,12.861,INTERMEDIARY
514,WpGC1178,"May 22, 2022",M,51.0,72.800,1.810000,22.200,NaN,126/82,101.0,...,Moderate,Y,181.000,0.508,126.0,82.0,Hypertension Stage 1,-7.000,12.760,INTERMEDIARY
528,uZes1835,2024-07-31,F,77.0,60.107,1.728000,22.247,89.011,166/116,108.0,...,Low,N,172.821,NaN,166.0,116.0,Hypertension Stage 2,-79.866,14.909,HIGH
666,OIoK9881,2024-07-20,F,44.0,101.500,1.690000,35.500,91.500,98/73,103.0,...,Moderate,N,169.000,0.541,98.0,73.0,Normal,-1.000,14.060,LOW
800,oduH3019,2023-08-13,F,55.0,82.400,1.600000,32.200,132.045,115/75,175.0,...,Moderate,N,160.000,NaN,115.0,75.0,Normal,-82.882,15.690,INTERMEDIARY
807,LPox8984,2024-08-18,F,26.0,66.394,1.860000,27.704,91.595,165/101,106.0,...,High,Y,185.957,0.493,165.0,101.0,Hypertension Stage 2,-2.000,16.964,LOW
987,tRLY5166,27 Jul 25,F,46.0,73.200,1.800000,22.600,79.300,126/81,135.0,...,Low,N,180.000,0.441,126.0,81.0,Hypertension Stage 1,-79.437,15.520,HIGH


In [146]:
datos.loc[datos.loc[:, 'Total Cholesterol (mg/dL)']<0]

,Patient ID,Date of Service,Sex,Age,Weight (kg),Height (m),BMI,Abdominal Circumference (cm),Blood Pressure (mmHg),Total Cholesterol (mg/dL),...,Physical Activity Level,Family History of CVD,Height (cm),Waist-to-Height Ratio,Systolic BP,Diastolic BP,Blood Pressure Category,Estimated LDL (mg/dL),CVD Risk Score,CVD Risk Level
1363,WVxJ0254,07-01-2020,F,70.0,58.35,1.542,23.761,100.617,95/110,-1.256,...,Moderate,Y,154.185,NaN,95.0,110.0,Hypertension Stage 2,64.0,12.122,INTERMEDIARY


In [147]:
cols_to_impute = ['Estimated LDL (mg/dL)', 'Total Cholesterol (mg/dL)', 'CVD Risk Score']

for col in cols_to_impute:
    negative_value = datos[datos[col] >= 0][col]*-1
    datos.loc[datos[col] < 0, col] = negative_value

print("Valores negativos imputados:")
for col in cols_to_impute:
    print(f"{col}: {(datos[col] < 0).sum()} valores negativos restantes")

Valores negativos imputados:
Estimated LDL (mg/dL): 0 valores negativos restantes
Total Cholesterol (mg/dL): 0 valores negativos restantes
CVD Risk Score: 0 valores negativos restantes


#### Agrupar registros repetidos

In [148]:
id_col = 'Patient ID'

num_cols = datos.select_dtypes(include='number').columns.tolist()

datos = (
    datos
    .groupby(id_col, as_index=False)[num_cols]
    .mean()
)

print(f"Número de pacientes únicos: {datos.shape[0]}")
datos.head()



Número de pacientes únicos: 1375


,Patient ID,Age,Weight (kg),Height (m),BMI,Abdominal Circumference (cm),Total Cholesterol (mg/dL),HDL (mg/dL),Fasting Blood Sugar (mg/dL),Height (cm),Waist-to-Height Ratio,Systolic BP,Diastolic BP,Estimated LDL (mg/dL),CVD Risk Score
0,AEFC1294,52.0,109.700,1.780,34.600,104.400,197.0,0.915,99.0,178.000,0.587,103.0,99.0,NaN,18.010
1,AHTL6366,NaN,104.469,1.995,25.390,86.894,NaN,82.000,175.0,199.458,0.436,164.0,90.0,147.0,18.458
2,AHjK2744,58.0,63.206,1.870,33.800,99.600,106.0,6.809,115.0,187.000,0.533,NaN,91.0,33.0,16.530
3,ALHn0227,52.0,107.800,1.780,34.000,103.500,179.0,73.000,132.0,178.000,0.581,111.0,69.0,76.0,15.930
4,AMxU2442,57.0,116.772,1.569,34.126,90.737,254.0,60.000,177.0,156.854,0.578,117.0,116.0,164.0,17.755


In [149]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1375 entries, 0 to 1374
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Patient ID                    1375 non-null   object 
 1   Age                           1316 non-null   float64
 2   Weight (kg)                   1375 non-null   float64
 3   Height (m)                    1375 non-null   float64
 4   BMI                           1332 non-null   float64
 5   Abdominal Circumference (cm)  1326 non-null   float64
 6   Total Cholesterol (mg/dL)     1317 non-null   float64
 7   HDL (mg/dL)                   1309 non-null   float64
 8   Fasting Blood Sugar (mg/dL)   1328 non-null   float64
 9   Height (cm)                   1375 non-null   float64
 10  Waist-to-Height Ratio         1312 non-null   float64
 11  Systolic BP                   1321 non-null   float64
 12  Diastolic BP                  1316 non-null   float64
 13  Est